### PR TITLE
002 view job spec

### PR DESCRIPTION
**Pull request by jedgar3**
Merged with Josh's approved files to keep things in the same place

### CODE REVIEW

File: src/main/java/org/kainos/ea/dao/DatabaseConnector.java [hunk 1]
```diff
@@ -7,22 +9,21 @@
...
 
         if (conn != null && !conn.isClosed()) {
             return conn;
         }
 
-        try(FileInputStream propStream = new FileInputStream("db.properties")) {
-
-            Properties props = new Properties();
-            props.load(propStream);
+        try {
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/dao/DatabaseConnector.java:20]: remove white space

File: src/main/java/org/kainos/ea/exception/DatabaseConnectionException.java
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/exception/DatabaseConnectionException.java:1]: Outstanding

File: src/main/java/org/kainos/ea/model/Capability.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+public class Capability {
+    private int id;
+    private String name;
+    private String description;
+
+    public Capability() {
+
+    }
+
+    public Capability(int id, String name, String description) {
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/Capability.java:12]: remove extra constructors and use only one in your code please

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package org.kainos.ea.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.io.Serializable;
+
+public class JobRole {
+
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/JobRole.java:9]: remove whitespace

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    public JobRole(int id, String name, Band band, Capability capability) {
+        this.setId(id);
+        this.setName(name);
+        this.setBand(band);
+        this.setCapability(capability);
+    }
+
+
+    public JobRole(int id, String name, Band band, String specifictaion) {
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/JobRole.java:27]: remove extra constructor and refactor code to just have one.

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.List;
+
+public class JobRoleService {
+
+    private JobRoleDao jobRoleDao;
+
+    public JobRoleService(JobRoleDao jobRoleDao){
+        this.jobRoleDao = jobRoleDao;
+    }
+
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/service/JobRoleService.java:18]: remove whitespace

File: src/test/java/unit/controller/JobRoleControllerUnitTest.java
Comment (joshuasy10) [inline src/test/java/unit/controller/JobRoleControllerUnitTest.java:1]: fantastic

**Review by joshuasy10 — CHANGES_REQUESTED** [review]
See comments, need to remove white space in places and remove extra constructors and refactor code to use just one.

**Review by joshuasy10 — APPROVED** [review]

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 6]
```diff
@@ -26,8 +27,26 @@ public Response getJobRoles() {
         try {
             return Response.ok(jobRoleService.getAllJobRoles()).build();
         } catch (FailedToGetJobRoleException e) {
+            System.err.println((e.getMessage()));
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        } catch (DatabaseConnectionException e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java:32]: bad request suggests that the user passed incorrect data in but they don't pass any data to this endpoint, please combine both catch blocks to return 500

File: src/main/java/org/kainos/ea/dao/JobRoleDao.java [hunk 7]
```diff
@@ -34,11 +35,42 @@ public List<JobRole> getAllJobRoles() throws SQLException {
                     rs.getInt("job_roles.id"),
                     rs.getString("job_roles.name"),
                     band,
-                    capability
+                    capability,
+                    null
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/JobRoleDao.java:39]: not sure why you need to pass null here? if you don't want to return the specification for this endpoint then create a new constructor that doesn't accept it or create a new response model that has a specification

File: src/main/java/org/kainos/ea/dao/JobRoleDao.java [hunk 8]
```diff
@@ -34,11 +35,42 @@ public List<JobRole> getAllJobRoles() throws SQLException {
...
+        while (rs.next()) {
+            Capability capability = new Capability(
+                    rs.getInt("capabilities.id"),
+                    rs.getString("capabilities.name"),
+                    rs.getString("capabilities.description")
+            );
+
+            Band band = new Band();
+            band.setName(rs.getString("bands.name"));
+            band.setLevel(rs.getString("level"));
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/JobRoleDao.java:64]: create a constructor to pass the through in 1 line

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 9]
```diff
@@ -1,24 +1,28 @@
...
-    public JobRole(int id, String name, Band band, Capability capability) {
-        this.setId(id);
-        this.setName(name);
-        this.setBand(band);
-        this.setCapability(capability);
+    @JsonCreator
+    public JobRole(@JsonProperty("id") int id,
+                   @JsonProperty("name") String name,
+                   @JsonProperty("band") Band band,
+                   @JsonProperty("Capability") Capability capability,
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/model/JobRole.java:17]:
```suggestion
                   @JsonProperty("capability") Capability capability,
```

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 10]
```diff
@@ -1,25 +1,39 @@
...
+    public JobRole getJobRolesById(int id) throws FailedToGetJobRoleException, DatabaseConnectionException {
+        try{
+            JobRole jobRole = jobRoleDao.getJobRoleById(id);
+
+            if (jobRole != null) {
+                return jobRole;
+            }
+            else{
+                throw new FailedToGetJobRoleException();
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/JobRoleService.java:32]: this should be a JobRoleDoesNotExist exception which should return a 404 from controller, FailedToGetJobRoleException should return a 500

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 11]
```diff
@@ -52,9 +51,21 @@ void getAllJobRoles_shouldReturnJobRoleList_whenDaoReturnsJobRoleList() throws S
...
     @Test
-    void getAllJobRoles_shouldThrowFailedToGetJobRoleException_whenDaoThrowsSQLException() throws SQLException{
+    void getAllJobRoles_shouldThrowFailedToGetJobRoleException_whenDaoThrowsSQLException() throws SQLException, DatabaseConnectionException {
         Mockito.when(jobRoleDao.getAllJobRoles()).thenThrow(SQLException.class);
 
         assertThrows(FailedToGetJobRoleException.class, () -> jobRoleService.getAllJobRoles());
     }
+
+    @Test
+    void getJobRolesById_shouldThrowFailedToGetJobRoleException_whenDaoReturnsNull() throws SQLException, DatabaseConnectionException {
```
Comment (shaunganley) [inline src/test/java/unit/service/JobRoleServiceTest.java:61]: need another unit test for SQLException being thrown

**Review by shaunganley — CHANGES_REQUESTED** [review]
good work, few small comments

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 12]
```diff
@@ -25,9 +27,24 @@ public JobRoleController(JobRoleService jobRoleService) {
     public Response getJobRoles() {
         try {
             return Response.ok(jobRoleService.getAllJobRoles()).build();
-        } catch (FailedToGetJobRoleException e) {
+        } catch (DatabaseConnectionException | FailedToGetJobRoleException e) {
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java:30]: Unbelievable stuff!

File: src/main/java/org/kainos/ea/exception/JobRoleDoesNotExistException.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package org.kainos.ea.exception;
+
+public class JobRoleDoesNotExistException extends Throwable {
+    @Override
+    public String getMessage(){
+        return "Job role does not exist";
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/exception/JobRoleDoesNotExistException.java:6]: Bloody fantastic!

File: src/main/java/org/kainos/ea/model/Band.java [hunk 14]
```diff
@@ -1,16 +1,32 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
 public class Band {
     private int id;
     private String name;
     private String level;
 
-    public Band(int id, String name, String level) {
+    public Band() {
+
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/Band.java:12]: Remove unneccessary constructor

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 15]
```diff
@@ -1,23 +1,34 @@
 package org.kainos.ea.model;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 
 public class JobRole {
-
-
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/JobRole.java:5]: amazing removal of white space!

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 16]
```diff
@@ -1,23 +1,34 @@
...
 
-    public JobRole(int id, String name, Band band, Capability capability) {
-        this.setId(id);
-        this.setName(name);
-        this.setBand(band);
-        this.setCapability(capability);
+    @JsonCreator
+    public JobRole(@JsonProperty("id") int id,
+                   @JsonProperty("name") String name,
+                   @JsonProperty("band") Band band,
```
Comment (joshuasy10) [inline src/main/java/org/kainos/ea/model/JobRole.java:16]: great debugging!

File: src/test/java/unit/service/JobRoleServiceTest.java [hunk 17]
```diff
@@ -36,11 +35,12 @@ public class JobRoleServiceTest {
             1,
             "Tim",
             band,
-            capability
+            capability,
```
Comment (joshuasy10) [inline src/test/java/unit/service/JobRoleServiceTest.java:38]: great stuff!

**Review by joshuasy10 — APPROVED** [review]

File: src/main/java/org/kainos/ea/model/Band.java [hunk 18]
```diff
@@ -1,16 +1,28 @@
...
+            @JsonProperty("level") String level) {
         this.setId(id);
         this.setName(name);
         this.setLevel(level);
     }
 
+    public Band(String name, String level) {
+        this.setName(name);
+        this.setLevel(level);
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/model/Band.java:24]: why do we need this constructor? can't we use the existing constructor and return the band ID?

**Review by shaunganley — APPROVED** [review]

**Pull request merged**